buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -439,40 +616,62 @@ private static class AllocationStatistics {
         private int histoIndex;
         private int datumCount;
         private int datumTarget = INIT_DATUM_TARGET;
-        protected boolean hasHadRotation;
-        protected volatile int sharedPrefChunkSize = MIN_CHUNK_SIZE;
-        protected volatile int localPrefChunkSize = MIN_CHUNK_SIZE;
-        protected volatile int localUpperBufSize;
+        private boolean hasHadRotation;
+        private volatile int sharedPrefChunkSize = MIN_CHUNK_SIZE;
+        private volatile int localPrefChunkSize = MIN_CHUNK_SIZE;
+        private volatile int localUpperBufSize;
 
-        private AllocationStatistics(AdaptivePoolingAllocator parent, boolean shareable) {
-            this.parent = parent;
+        private HistogramChunkController(MagazineGroup group, boolean shareable) {
+            this.group = group;
             this.shareable = shareable;
         }
 
-        protected void recordAllocationSize(int bufferSizeToRecord) {
+        @Override
+        public int computeBufferCapacity(
+                int requestedSize, int maxCapacity, boolean isReallocation) {
+            if (!isReallocation) {
+                // Only record allocation size if it's not caused by a reallocation that was triggered by capacity
+                // change of the buffer.
+                recordAllocationSize(requestedSize);
+            }
+
+            // Predict starting capacity from localUpperBufSize, but place limits on the max starting capacity
+            // based on the requested size, because localUpperBufSize can potentially be quite large.
+            int startCapLimits;
+            if (requestedSize <= 32768) { // Less than or equal to 32 KiB.
+                startCapLimits = 65536; // Use at most 64 KiB, which is also the AdaptiveRecvByteBufAllocator max.
```

Comment:
just a comment... while we use 64KiB as max for `AdaptiveRecvByteBufAllocator` by default I wonder for some time if this is really a good default these days.

Reply:
Yeah, I'm thinking we should get rid of these heuristics.